primitives/bridge/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+            let command = queued_message.command.index();
+            let params = queued_message.command.abi_encode();
+            let max_dispatch_gas =
+                ConstantGasMeter::maximum_dispatch_gas_used_at_most(&queued_message.command);
+            let reward = pricing_params.rewards.remote;
+
+            // Construct the final committed message
+            let message = CommittedMessage {
+                channel_id: queued_message.channel_id,
+                nonce,
+                command,
+                params,
+                max_dispatch_gas,
+                max_fee_per_gas: pricing_params
+                    .fee_per_gas
+                    .try_into()
+                    .defensive_unwrap_or(u128::MAX),
+                reward: reward.try_into().defensive_unwrap_or(u128::MAX),
+                id: queued_message.id,
+            };
+
+            // ABI-encode and hash the prepared message
+            let message_abi_encoded = ethabi::encode(&[message.clone().into()]);
+            let message_abi_encoded_hash =
+                <T as snowbridge_pallet_outbound_queue::Config>::Hashing::hash(
+                    &message_abi_encoded,
+                );
+
+            Messages::<T>::append(Box::new(message));
+            MessageLeaves::<T>::append(message_abi_encoded_hash);
+
+            snowbridge_pallet_outbound_queue::Pallet::<T>::deposit_event(
+                snowbridge_pallet_outbound_queue::Event::MessageAccepted {
+                    id: queued_message.id,
+                    nonce,
+                },
+            );
+
+            Ok(true)
+        }
+    }
+
+    impl<T> ProcessMessage for CustomProcessSnowbridgeMessage<T>
+    where
+        T: snowbridge_pallet_outbound_queue::Config,
+    {
+        type Origin = T::AggregateMessageOrigin;
+
+        fn process_message(
+            message: &[u8],
+            origin: Self::Origin,
+            meter: &mut WeightMeter,
+            id: &mut [u8; 32],
+        ) -> Result<bool, ProcessMessageError> {
+            // TODO: this weight is from the pallet, should be very similar to the weight of
+            // Self::do_process_message, but ideally we should benchmark this separately
+            let original_meter = meter.clone();
+            let weight = T::WeightInfo::do_process_message();
+            if meter.try_consume(weight).is_err() {
+                return Err(ProcessMessageError::Overweight(weight));
+            }
+
+            match Self::do_process_message(origin.clone(), message) {
+                Ok(x) => Ok(x),
+                Err(ProcessMessageError::Corrupt) => {
+                    // In case of decode error, this may be a command from the original pallet.
+                    // So attempt to use the original `process_message` impl.
+                    // Refund weight to avoid charging double for snowbridge messages, `process_message` will consume the weight again.
```

Comment:
are we doing the refund? based of this would it be easier to simply put a different origin than `snowbridge`?